![alt text](https://raw.githubusercontent.com/collaborativebioinformatics/STRdust/main/STRdust-logo.jpg)

## Contributors

Anneri LÃ¶tter, Susanne P. Pfeifer - Writers

Luis Paulin

Damaris Lattimer

Deepak Choubey

Guangyi Chen

Kyuil Cho - Sysadmin

Kimberley Billingsley

Pavel Avdeyev

Simone Cree

Wouter De Coster - Team Lead

Yilei Fu

## Goals
* To develop a tool to detect and genotype (in terms of length) STR in long reads (_de novo_) without the need of genome annotation beforehand
* This tool should be applicable in mammals and plants (basically any phased eukaryotic assembly)

## Introduction/Description
Short tandem repeats are motifs of multiple nucleotides that are repeated. Long-read sequencing can be used to identify their length more accurately than short reads as reads can span across the entire repeat region. This tools makes use of phased read alignments to find and genotype STR.

## How does it work?

![alt text](https://raw.githubusercontent.com/collaborativebioinformatics/STR_Integration/main/Flow%20chart%20group2.jpg)

## How to use?

## Quickstart

### Input

### Output

## Testing

## Installation

 
